src/main/java/org/springframework/data/domain/KeysetScrollPosition.java

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2023 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.data.domain;
+
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+import org.springframework.util.Assert;
+import org.springframework.util.ObjectUtils;
+
+/**
+ * A {@link ScrollPosition} based on the last seen keyset. Keyset scrolling must be associated with a {@link Sort
+ * well-defined sort} to be able to extract the keyset when resuming scrolling within the sorted result set.
+ *
+ * @author Mark Paluch
+ * @since 3.1
+ */
+public final class KeysetScrollPosition implements ScrollPosition {
+
+	private static final KeysetScrollPosition initial = new KeysetScrollPosition(Collections.emptyMap(),
+			Direction.Forward);
+
+	private final Map<String, Object> keys;
+
+	private final Direction direction;
+
+	private KeysetScrollPosition(Map<String, Object> keys, Direction direction) {
+		this.keys = keys;
+		this.direction = direction;
+	}
+
+	/**
+	 * Creates a new initial {@link KeysetScrollPosition} to start scrolling using keyset-queries.
+	 *
+	 * @return a new initial {@link KeysetScrollPosition} to start scrolling using keyset-queries.
+	 */
+	public static KeysetScrollPosition initial() {
+		return initial;
+	}
+
+	/**
+	 * Creates a new {@link KeysetScrollPosition} from a keyset.
+	 *
+	 * @param keys must not be {@literal null}.
+	 * @return a new {@link KeysetScrollPosition} for the given keyset.
+	 */
+	public static KeysetScrollPosition of(Map<String, ?> keys) {
+		return of(keys, Direction.Forward);
+	}
+
+	/**
+	 * Creates a new {@link KeysetScrollPosition} from a keyset and {@link Direction}.
+	 *
+	 * @param keys must not be {@literal null}.
+	 * @param direction must not be {@literal null}.
+	 * @return a new {@link KeysetScrollPosition} for the given keyset and {@link Direction}.
+	 */
+	public static KeysetScrollPosition of(Map<String, ?> keys, Direction direction) {
```

Comment:
This is impossible to use currently since `Direction` is declared package private.